scripts/usb_scan.py

```diff
@@ -0,0 +1,725 @@
+                usb = True
+        return None
+
+    def is_child_of(self, parent):
+        if type(parent) is UsbInterface and parent.is_initialized():
+            return parent.get_node() == self.get_parent_node()
+        return False
+
+
+class Singleton(type):
+    _instances = {}
+
+    def __call__(cls, *args, **kwargs):
+        if cls not in cls._instances:
+            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
+        return cls._instances[cls]
+
+
+class Scanner(object):
+    """ get USB info from UDEV database
+
+    """
+    __metaclass__ = Singleton
+
+    def __init__(self):
+        self.context = pyudev.Context()
+
+    def get_usb_devices(self):
+        ret = []
+        for d in self.context.list_devices(subsystem="usb", DEVTYPE="usb_device"):
+            device = UsbDevice(d.sys_name, d, d.attributes)
+            if device.is_initialized() and not device.is_class_hub():
+                ret.append(device)
+            else:
+                log_wrapper("ignore usb device:" + str(device))
+        return ret
+
+    def get_usb_interfaces(self):
+        ret = []
+        for d in self.context.list_devices(subsystem="usb", DEVTYPE="usb_interface"):
+            interface = UsbInterface(d.sys_name, d.attributes)
+            if interface.is_initialized() and not interface.is_class_hub():
+                ret.append(interface)
+            else:
+                log_wrapper("ignore usb interface:" + str(interface))
+        return ret
+
+    def get_usb_storages(self):
+        ret = []
+        for d in self.context.list_devices(subsystem="block", DEVTYPE="disk"):
+            if d.sys_path is None or d.sys_path.find("usb") == -1:
+                continue
+            storage = UsbStorage(d.sys_name, d.sys_path)
+            if storage.is_initialized():
+                ret.append(storage)
+            else:
+                log_wrapper("ignore usb storage:" + str(storage))
+        return ret
+
+
+class Policy(object):
+    """ Parse policy file, and check if a UsbDevice can be passed through
+
+    Policy file spec reference:
+    https://support.citrix.com/article/CTX119722
+
+    rule_list: the list of parsed rule
+    """
+    __metaclass__ = Singleton
+
+    _PATH = "/etc/xensource/usb-policy.conf"
+
+    _CLASS = "class"
+    _SUBCLASS = "subclass"
+    _PROTOCOL = "prot"
+    _ID_VENDOR = "vid"
+    _ID_PRODUCT = "pid"
+    _BCD_DEVICE = "rel"
+
+    # key in policy <--> key in usb device
+    _KEY_MAP_DEVICE = {_ID_VENDOR: UsbDevice._ID_VENDOR, _ID_PRODUCT: UsbDevice._ID_PRODUCT,
+                       _BCD_DEVICE: UsbDevice._BCD_DEVICE}
+
+    # key in policy <--> key in usb interface
+    _KEY_MAP_INTERFACE = {_CLASS: UsbInterface._CLASS, _SUBCLASS: UsbInterface._SUB_CLASS,
+                          _PROTOCOL: UsbInterface._PROTOCOL}
+
+    _PAT_KEY = r"\s*({}|{}|{}|{}|{}|{})\s*".format(_CLASS, _SUBCLASS, _PROTOCOL, _ID_VENDOR, _ID_PRODUCT,
```

Comment:
Some of these lines look longer than 80 characters, please run `pep8` on this file.

Reply:
OK